File: Fuel Consumption and Efficiency Analysis Project/scripts/python/statistical_analysis.py
"""
Statistical Analysis Script
Performs descriptive, inferential, and exploratory statistical analysis
"""

import pandas as pd
import numpy as np
from scipy import stats
from scipy.stats import normaltest, shapiro, pearsonr, spearmanr
import os
import sys

sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

def load_data():
    """Load the dataset"""
    data_path = os.path.join(os.path.dirname(__file__), '..', '..', 'data', 'FuelConsumption.csv')
    df = pd.read_csv(data_path)
    df.columns = df.columns.str.strip()
    return df

def descriptive_statistics(df):
    """Calculate descriptive statistics"""
    print("="*50)
    print("DESCRIPTIVE STATISTICS")
    print("="*50)
    
    numerical_cols = ['ENGINE SIZE', 'CYLINDERS', 'FUEL CONSUMPTION', 'COEMISSIONS']
    
    descriptive_stats = pd.DataFrame({
        'Mean': df[numerical_cols].mean(),
        'Median': df[numerical_cols].median(),
        'Std': df[numerical_cols].std(),
        'Variance': df[numerical_cols].var(),
        'Min': df[numerical_cols].min(),
        'Max': df[numerical_cols].max(),
        'Range': df[numerical_cols].max() - df[numerical_cols].min(),
        'Skewness': df[numerical_cols].skew(),
        'Kurtosis': df[numerical_cols].kurtosis()
    })
    
    print("\nDescriptive Statistics:")
    print(descriptive_stats)
    
    print("\nQuartiles and IQR:")
    for col in numerical_cols:
        Q1 = df[col].quantile(0.25)
        Q2 = df[col].quantile(0.50)
        Q3 = df[col].quantile(0.75)
        IQR = Q3 - Q1
        print(f"\n{col}:")
        print(f"  Q1: {Q1:.2f}, Q2: {Q2:.2f}, Q3: {Q3:.2f}")
        print(f"  IQR: {IQR:.2f}")

def inferential_statistics(df):
    """Perform inferential statistics"""
    print("\n" + "="*50)
    print("INFERENTIAL STATISTICS")
    print("="*50)
    
    numerical_cols = ['ENGINE SIZE', 'CYLINDERS', 'FUEL CONSUMPTION', 'COEMISSIONS']
    
    print("\nNormality Tests:")
    for col in numerical_cols:
        data = df[col].dropna()
        if len(data) < 5000:
            stat, p_value = shapiro(data)
            test_name = "Shapiro-Wilk"
        else:
            stat, p_value = normaltest(data)
            test_name = "D'Agostino"
        
        print(f"\n{col} ({test_name}):")
        print(f"  Statistic: {stat:.4f}, p-value: {p_value:.4f}")
        print(f"  Normal: {'Yes' if p_value > 0.05 else 'No'}")

def exploratory_statistics(df):
    """Perform exploratory statistical analysis"""
    print("\n" + "="*50)
    print("EXPLORATORY STATISTICAL ANALYSIS")
    print("="*50)
    
    numerical_cols = ['ENGINE SIZE', 'CYLINDERS', 'FUEL CONSUMPTION', 'COEMISSIONS']
    
    print("\n95% Confidence Intervals for Mean:")
    for col in numerical_cols:
        data = df[col].dropna()
        mean = data.mean()
        n = len(data)
        ci = stats.t.interval(0.95, n-1, loc=mean, scale=stats.sem(data))
        print(f"\n{col}:")
        print(f"  Mean: {mean:.2f}")
        print(f"  95% CI: [{ci[0]:.2f}, {ci[1]:.2f}]")
    
    print("\nCorrelation Analysis with Significance:")
    target = 'COEMISSIONS'
    for col in ['ENGINE SIZE', 'CYLINDERS', 'FUEL CONSUMPTION']:
        data = df[[col, target]].dropna()
        pearson_corr, pearson_p = pearsonr(data[col], data[target])
        spearman_corr, spearman_p = spearmanr(data[col], data[target])
        print(f"\n{col} vs {target}:")
        print(f"  Pearson: r={pearson_corr:.4f}, p={pearson_p:.4f}")
        print(f"  Spearman: ρ={spearman_corr:.4f}, p={spearman_p:.4f}")

def main():
    """Main function"""
    df = load_data()
    descriptive_statistics(df)
    inferential_statistics(df)
    exploratory_statistics(df)

if __name__ == "__main__":
    main()

File: Fuel Consumption and Efficiency Analysis Project/scripts/r/statistical_analysis.R
# Statistical Analysis Script
# Performs descriptive, inferential, and exploratory statistical analysis

# Load necessary libraries
library(dplyr)
library(psych)
library(car)

# Suppress warnings
options(warn = -1)

# Load the dataset
data_path <- file.path("..", "..", "data", "FuelConsumption.csv")
df <- read.csv(data_path, stringsAsFactors = FALSE)

# Clean column names
colnames(df) <- trimws(colnames(df))

cat("==================================================\n")
cat("STATISTICAL ANALYSIS\n")
cat("==================================================\n\n")

# Descriptive Statistics
cat("==================================================\n")
cat("DESCRIPTIVE STATISTICS\n")
cat("==================================================\n\n")

numerical_cols <- c("ENGINE.SIZE", "CYLINDERS", "FUEL.CONSUMPTION", "COEMISSIONS")

# Calculate descriptive statistics
descriptive_stats <- df %>%
  select(all_of(numerical_cols)) %>%
  describe() %>%
  select(mean, median, sd, var, min, max, skew, kurtosis)

cat("Descriptive Statistics:\n")
print(round(descriptive_stats, 3))

# Quartiles and IQR
cat("\nQuartiles and IQR:\n")
for(col in numerical_cols) {
  Q1 <- quantile(df[[col]], 0.25, na.rm = TRUE)
  Q2 <- quantile(df[[col]], 0.50, na.rm = TRUE)
  Q3 <- quantile(df[[col]], 0.75, na.rm = TRUE)
  IQR <- Q3 - Q1
  cat("\n", col, ":\n")
  cat("  Q1:", round(Q1, 2), ", Q2:", round(Q2, 2), ", Q3:", round(Q3, 2), "\n")
  cat("  IQR:", round(IQR, 2), "\n")
}

# Inferential Statistics
cat("\n==================================================\n")
cat("INFERENTIAL STATISTICS\n")
cat("==================================================\n\n")

# Normality Tests
cat("Normality Tests:\n")
for(col in numerical_cols) {
  data <- df[[col]][!is.na(df[[col]])]
  if(length(data) < 5000) {
    test_result <- shapiro.test(data)
    test_name <- "Shapiro-Wilk"
    stat <- test_result$statistic
    p_value <- test_result$p.value
  } else {
    test_result <- ks.test(data, "pnorm", mean(data), sd(data))
    test_name <- "Kolmogorov-Smirnov"
    stat <- test_result$statistic
    p_value <- test_result$p.value
  }
  
  cat("\n", col, " (", test_name, "):\n")
  cat("  Statistic:", round(stat, 4), ", p-value:", round(p_value, 4), "\n")
  cat("  Normal:", ifelse(p_value > 0.05, "Yes", "No"), "\n")
}

# T-tests: Compare fuel consumption by fuel type
cat("\nT-test: Fuel Consumption by Fuel Type\n")
fuel_types <- unique(df$FUEL)
if(length(fuel_types) >= 2) {
  group1 <- df[df$FUEL == fuel_types[1], "FUEL.CONSUMPTION"]
  group2 <- df[df$FUEL == fuel_types[2], "FUEL.CONSUMPTION"]
  t_test <- t.test(group1, group2)
  cat("  T-statistic:", round(t_test$statistic, 4), "\n")
  cat("  P-value:", round(t_test$p.value, 4), "\n")
  cat("  Significant difference:", ifelse(t_test$p.value < 0.05, "Yes", "No"), "\n")
}

# ANOVA: Compare fuel consumption across vehicle classes
cat("\nANOVA: Fuel Consumption by Vehicle Class (Top 5)\n")
top_classes <- df %>%
  count(VEHICLE.CLASS, sort = TRUE) %>%
  head(5) %>%
  pull(VEHICLE.CLASS)

groups <- lapply(top_classes, function(x) {
  df[df$VEHICLE.CLASS == x, "FUEL.CONSUMPTION"]
})

anova_result <- aov(FUEL.CONSUMPTION ~ VEHICLE.CLASS, 
                    data = df[df$VEHICLE.CLASS %in% top_classes, ])
anova_summary <- summary(anova_result)
cat("  F-statistic:", round(anova_summary[[1]][["F value"]][1], 4), "\n")
cat("  P-value:", round(anova_summary[[1]][["Pr(>F)"]][1], 4), "\n")
cat("  Significant difference:", 
    ifelse(anova_summary[[1]][["Pr(>F)"]][1] < 0.05, "Yes", "No"), "\n")

# Exploratory Statistical Analysis
cat("\n==================================================\n")
cat("EXPLORATORY STATISTICAL ANALYSIS\n")
cat("==================================================\n\n")

# Confidence Intervals
cat("95% Confidence Intervals for Mean:\n")
for(col in numerical_cols) {
  data <- df[[col]][!is.na(df[[col]])]
  mean_val <- mean(data)
  n <- length(data)
  se <- sd(data) / sqrt(n)
  ci_lower <- mean_val - qt(0.975, n-1) * se
  ci_upper <- mean_val + qt(0.975, n-1) * se
  
  cat("\n", col, ":\n")
  cat("  Mean:", round(mean_val, 2), "\n")
  cat("  95% CI: [", round(ci_lower, 2), ", ", round(ci_upper, 2), "]\n")
}

# Correlation Analysis with p-values
cat("\nCorrelation Analysis with Significance:\n")
target <- "COEMISSIONS"
for(col in c("ENGINE.SIZE", "CYLINDERS", "FUEL.CONSUMPTION")) {
  data_subset <- df[, c(col, target)]
  data_subset <- data_subset[complete.cases(data_subset), ]
  
  pearson_test <- cor.test(data_subset[[col]], data_subset[[target]], 
                           method = "pearson")
  spearman_test <- cor.test(data_subset[[col]], data_subset[[target]], 
                            method = "spearman")
  
  cat("\n", col, " vs ", target, ":\n")
  cat("  Pearson: r =", round(pearson_test$estimate, 4), 
      ", p =", round(pearson_test$p.value, 4), "\n")
  cat("  Spearman: ρ =", round(spearman_test$estimate, 4), 
      ", p =", round(spearman_test$p.value, 4), "\n")
  cat("  Significant:", ifelse(pearson_test$p.value < 0.05, "Yes", "No"), "\n")
}

cat("\n==================================================\n")
cat("STATISTICAL ANALYSIS COMPLETE!\n")
cat("==================================================\n")
